shell/platform/windows/text_input_plugin.cc

```diff
@@ -165,23 +185,41 @@ void TextInputPlugin::HandleMethodCall(
                     "Set editing state has been invoked, but without text.");
       return;
     }
-    auto selection_base = args.FindMember(kSelectionBaseKey);
-    auto selection_extent = args.FindMember(kSelectionExtentKey);
-    if (selection_base == args.MemberEnd() || selection_base->value.IsNull() ||
-        selection_extent == args.MemberEnd() ||
-        selection_extent->value.IsNull()) {
+    auto base = args.FindMember(kSelectionBaseKey);
+    auto extent = args.FindMember(kSelectionExtentKey);
+    if (base == args.MemberEnd() || base->value.IsNull() ||
+        extent == args.MemberEnd() || extent->value.IsNull()) {
       result->Error(kInternalConsistencyError,
                     "Selection base/extent values invalid.");
       return;
     }
     // Flutter uses -1/-1 for invalid; translate that to 0/0 for the model.
-    int base = selection_base->value.GetInt();
-    int extent = selection_extent->value.GetInt();
-    if (base == -1 && extent == -1) {
-      base = extent = 0;
+    int selection_base = base->value.GetInt();
+    int selection_extent = extent->value.GetInt();
+    if (selection_base == -1 && selection_extent == -1) {
+      selection_base = selection_extent = 0;
     }
     active_model_->SetText(text->value.GetString());
-    active_model_->SetSelection(TextRange(base, extent));
+    active_model_->SetSelection(TextRange(selection_base, selection_extent));
+
+    base = args.FindMember(kComposingBaseKey);
+    extent = args.FindMember(kComposingExtentKey);
+    if (base == args.MemberEnd() || base->value.IsNull() ||
+        extent == args.MemberEnd() || extent->value.IsNull()) {
+      result->Error(kInternalConsistencyError,
+                    "Composing base/extent values invalid.");
+      return;
+    }
+    int composing_base = base->value.GetInt();
+    int composing_extent = base->value.GetInt();
+    if (composing_base == -1 && composing_extent == -1) {
+      active_model_->EndComposing();
+    } else {
+      int composing_start = std::min(composing_base, composing_extent);
+      int cursor_offset = selection_base - composing_start;
+      active_model_->SetComposingRange(
+          TextRange(composing_base, composing_extent), cursor_offset);
+    }
```

Comment:
It would be nice to reduce duplication here with a private range class and a helper to populate it from a dictionary (given the keys as arguments). Up to you if it's worth doing though.